### PR TITLE
Add support for Material 3 medium and large top app bars.

### LINKED ISSUE #102684: Update AppBar to support new M3 'Medium' and 'Large' variants.
While #101884 added base [Material 3](https://m3.material.io/components/top-app-bar/overview) support for `AppBar` there are two new variants that are not yet supported: [Medium](https://m3.material.io/components/top-app-bar/specs#e3fd3eba-0444-437c-9a82-071ef03d85b1) and [Large](https://m3.material.io/components/top-app-bar/specs#8140aaaf-5729-4368-a0f5-baef8d576dbf):

Screen Shot 2022-04-27 at 12 51 50 PM

Which also provide a scroll under animation when they are collapsed.

### PR DESCRIPTION
**Pull request by darrenaustin**
A follow on to #101884.

This adds support for the new Material 3 'Medium' and 'Large' type app bars.

 

 

Two new factory methods for `SliverAppBar` have been added: `SliverAppBar.medium` and `SliverAppBar.large`. These can be used inside a `CustomScrollView` like other `SliverAppBar`s, but will be configured for the 'Medium' or 'Large' defaults and behaviour.

Here is an example of how to create a 'Medium' app bar:

```dart
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(useMaterial3: true, colorSchemeSeed: const Color(0xff6750A4)),
      home: Material(
        child: CustomScrollView(
          slivers: <Widget>[
            SliverAppBar.medium(
              leading: IconButton(icon: const Icon(Icons.menu), onPressed: () {}),
              title: const Text('Medium App Bar'),
              actions: <Widget>[
                IconButton(icon: const Icon(Icons.more_vert), onPressed: () {}),
              ],
            ),

            // The rest of the scrolling slivers...
 ```

Fixes: #102684.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: examples/api/lib/material/app_bar/sliver_app_bar.3.dart [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  runApp(const AppBarLargeApp());
+}
+
+class AppBarLargeApp extends StatelessWidget {
+  const AppBarLargeApp({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      theme: ThemeData(useMaterial3: true, colorSchemeSeed: const Color(0xff6750A4)),
```
Comment (guidezpl) [inline examples/api/lib/material/app_bar/sliver_app_bar.3.dart:19]:
```suggestion
      theme: ThemeData(
        useMaterial3: true, 
        colorSchemeSeed: const Color(0xff6750A4),
      ),
```

File: examples/api/lib/material/app_bar/sliver_app_bar.3.dart [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+              ],
+            ),
+            // Just some content big enough to have something to scroll.
+            SliverToBoxAdapter(
+              child: Card(
+                child: SizedBox(
+                  height: 1200,
+                  child: Padding(
+                    padding: const EdgeInsets.fromLTRB(8, 100, 8, 100),
+                    child: Text('Here be scrolling content...', style: Theme.of(context).textTheme.headlineSmall),
```
Comment (guidezpl) [inline examples/api/lib/material/app_bar/sliver_app_bar.3.dart:37]:
```suggestion
                    child: Text(
                      'Here be scrolling content...', 
                      style: Theme.of(context).textTheme.headlineSmall,
                    ),
```

File: packages/flutter/lib/src/material/app_bar.dart [hunk 3]
```diff
@@ -1496,6 +1502,203 @@ class SliverAppBar extends StatefulWidget {
...
+      shape: shape,
+      toolbarHeight: toolbarHeight,
+      leadingWidth: leadingWidth,
+      toolbarTextStyle: toolbarTextStyle,
+      titleTextStyle: titleTextStyle,
+      systemOverlayStyle: systemOverlayStyle,
+    );
+  }
+
+  /// Creates a Material Design app bar matching the 'Large' type
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/app_bar.dart:1603]:
```suggestion
  /// Creates a large Material Design top app bar
```

File: packages/flutter/lib/src/material/app_bar.dart [hunk 4]
```diff
@@ -1496,6 +1502,203 @@ class SliverAppBar extends StatefulWidget {
        assert(stretchTriggerOffset > 0.0),
        assert(collapsedHeight == null || collapsedHeight >= toolbarHeight, 'The "collapsedHeight" argument has to be larger than or equal to [toolbarHeight].');
 
+  /// Creates a Material Design app bar matching the 'Medium' type
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/app_bar.dart:1505]:
```suggestion
  /// Creates a large Material Design top app bar
```

File: packages/flutter/lib/src/material/app_bar.dart [hunk 5]
```diff
@@ -1496,6 +1502,203 @@ class SliverAppBar extends StatefulWidget {
...
+  /// that can be placed in a [CustomScrollView].
+  ///
+  /// By default, it configures a [SliverAppBar] with the appropriate defaults
+  /// for the 'Medium' type Material 3 app bar. It starts fully expanded with
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/app_bar.dart:1509]:
```suggestion
  /// By default, configures a [SliverAppBar] with the appropriate defaults
  /// for a medium Material 3 top app bar. It starts fully expanded with
```

File: packages/flutter/lib/src/material/app_bar.dart [hunk 6]
```diff
@@ -1496,6 +1502,203 @@ class SliverAppBar extends StatefulWidget {
...
+      titleTextStyle: titleTextStyle,
+      systemOverlayStyle: systemOverlayStyle,
+    );
+  }
+
+  /// Creates a Material Design app bar matching the 'Large' type
+  /// that can be placed in a [CustomScrollView].
+  ///
+  /// By default, it configures a [SliverAppBar] with the appropriate defaults
+  /// for the 'Large' type Material 3 app bar. It starts fully expanded with
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/app_bar.dart:1607]:
```suggestion
  /// By default, configures a [SliverAppBar] with the appropriate defaults
  /// for a large Material 3 top app bar. It starts fully expanded with
```
```

File: packages/flutter/lib/src/material/app_bar.dart [hunk 7]
```diff
@@ -841,8 +841,9 @@ class _AppBarState extends State<AppBar> {
       ?? MaterialStateProperty.resolveAs<Color>(defaultColor, states);
   }
 
-  SystemUiOverlayStyle _systemOverlayStyleForBrightness(Brightness brightness) {
-    return brightness == Brightness.dark ? SystemUiOverlayStyle.light : SystemUiOverlayStyle.dark;
+  SystemUiOverlayStyle _systemOverlayStyleForBrightness(Brightness brightness, [Color? backgroundColor]) {
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/app_bar.dart:844]: nit: formatting

File: packages/flutter/lib/src/material/app_bar.dart [hunk 8]
```diff
@@ -1496,6 +1502,203 @@ class SliverAppBar extends StatefulWidget {
...
+  /// {@tool dartpad}
+  /// This sample shows how to use [SliverAppBar.medium] in a [CustomScrollView].
+  ///
+  /// ** See code in examples/api/lib/material/app_bar/sliver_app_bar.2.dart **
+  /// {@end-tool}
+  ///
+  /// See also:
+  ///
+  ///   * [AppBar], for the 'Small' and 'Center-aligned' types of app bar.
+  ///   * [SliverAppBar.large], for the 'Large' type of app bar.
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/app_bar.dart:1525]:
```suggestion
  ///   * [AppBar], for a small or center-aligned top app bar.
  ///   * [SliverAppBar.large], for a large top app bar.
```

File: packages/flutter/lib/src/material/app_bar.dart [hunk 9]
```diff
@@ -1496,6 +1502,203 @@ class SliverAppBar extends StatefulWidget {
...
+  /// {@tool dartpad}
+  /// This sample shows how to use [SliverAppBar.large] in a [CustomScrollView].
+  ///
+  /// ** See code in examples/api/lib/material/app_bar/sliver_app_bar.3.dart **
+  /// {@end-tool}
+  ///
+  /// See also:
+  ///
+  ///   * [AppBar], for the 'Small' and 'Center-aligned' types of app bar.
+  ///   * [SliverAppBar.medium], for the 'Medium' type of app bar.
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/app_bar.dart:1623]:
```suggestion
  ///   * [AppBar], for a small or center-aligned top app bar.
  ///   * [SliverAppBar.medium], for a medium top app bar.
```

File: packages/flutter/lib/src/material/app_bar.dart [hunk 10]
```diff
@@ -1496,6 +1502,203 @@ class SliverAppBar extends StatefulWidget {
...
+    TextStyle? titleTextStyle,
+    SystemUiOverlayStyle? systemOverlayStyle,
+  }) {
+    return SliverAppBar(
+      key: key,
+      leading: leading,
+      automaticallyImplyLeading: automaticallyImplyLeading,
+      actions: actions,
+      flexibleSpace: flexibleSpace ?? _ScrollUnderFlexibleSpace(
+        title: title,
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/app_bar.dart:1569]: missing `variant: _ScrollUnderFlexibleVariant.medium,`?

**Review by guidezpl — COMMENTED** [review]
Nice, love the screen recordings. It's unfortunate to have to redefine long argument lists to modify a few defaults. Could this be improved by collapsing the two factory constructors into one, with an additional parameter specifying the size?

**Comment by darrenaustin** [comment]
>  It's unfortunate to have to redefine long argument lists to modify a few defaults. Could this be improved by collapsing the two factory constructors into one, with an additional parameter specifying the size?

I thought about it, but the naming would get a bit weird (i.e. what is the name of the single constructor? `mediumOrLarge`, or `scrolledUnder`?). And then we need to either add an awkward `mediumSize` bool or introducing a new public enum just for this one parameter.

While it has duplication in the parameter list, it still seemed like the easiest to understand API. If others disagree with this, I can certainly take a stab at combining them.

File: packages/flutter/lib/src/material/app_bar.dart (hunk 10 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/app_bar.dart:1571]: This was the default, so wasn't needed (and we have a lint that would complain if we specified it). But to make this more obvious, I removed the default and made it a required param. Thx.

**Review by guidezpl — CHANGES_REQUESTED** [review]
~Can we add Dartpad(s) like https://github.com/flutter/flutter/pull/102823?~

Ignore me, just saw them :)

File: packages/flutter/lib/src/material/app_bar.dart [hunk 11]
```diff
@@ -1496,6 +1504,204 @@ class SliverAppBar extends StatefulWidget {
...
+  /// ** See code in examples/api/lib/material/app_bar/sliver_app_bar.2.dart **
+  /// {@end-tool}
+  ///
+  /// See also:
+  ///
+  ///   * [AppBar], for a small or center-aligned top app bar.
+  ///   * [SliverAppBar.large], for a large top app bar.
+  ///   * https://m3.material.io/components/top-app-bar/overview, the Material 3
+  ///     app bar specification.
+  factory SliverAppBar.medium({
```
Comment (Piinks) [inline packages/flutter/lib/src/material/app_bar.dart:1530]: Bless you for omitting deprecated parameters. ❤️ 

Is `forceElevated` intentionally left out?

**Review by Piinks — COMMENTED** [review]
@darrenaustin this is awesome!

File: packages/flutter/lib/src/material/app_bar.dart [hunk 12]
```diff
@@ -1922,6 +2128,128 @@ class _RenderAppBarTitleBox extends RenderAligningShiftedBox {
...
+      expandedTitle = config.expandedTextStyle != null
+        ? DefaultTextStyle(
+            style: config.expandedTextStyle!,
+            child: widget.title!,
+          )
+        : widget.title;
+    }
+
+    late final bool centerTitle;
+    {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/app_bar.dart:2186]: NICE. I like blocks.

**Review by HansMuller — APPROVED** [review]
LGTM, just some trivial feedback.

**Review by guidezpl — APPROVED** [review]

**Comment by cedvdb** [comment]
It would be user friendly if the scaffold API incorporated those without having to pass through a CustomScrollView.

File: packages/flutter/lib/src/material/app_bar.dart (hunk 11 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/app_bar.dart:1530]: > Is `forceElevated` intentionally left out?

Somewhat, as it is kind of redundant now that we have two elevations (`elevation` & `scrolledUnderElevation`). But for people that don't want to specify both, and to be consistent with the other APIs perhaps I should add it.

File: packages/flutter/lib/src/material/app_bar.dart [hunk 13]
```diff
@@ -1496,6 +1504,204 @@ class SliverAppBar extends StatefulWidget {
...
+    Color? surfaceTintColor,
+    Color? backgroundColor,
+    Color? foregroundColor,
+    IconThemeData? iconTheme,
+    IconThemeData? actionsIconTheme,
+    bool primary = true,
+    bool? centerTitle,
+    bool excludeHeaderSemantics = false,
+    double? titleSpacing,
+    double collapsedHeight = _MediumScrollUnderFlexibleConfig.collapsedHeight,
```
[the remark this answers is not in the document]
Comment (darrenaustin) [reply packages/flutter/lib/src/material/app_bar.dart]: Good call. Thx.

File: packages/flutter/lib/src/material/app_bar.dart (hunk 12 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/app_bar.dart:2211]: Yeah, I don't like polluting the namespace with functions that are only needed one place. In cases like this though I really wish Dart had some kind of 'switch expression' that could be used for cases where we just need a value based on an enum.

**Comment by darrenaustin** [comment]
> It would be user friendly if the scaffold API incorporated those without having to pass through a CustomScrollView.

It would be nice, but I am not sure how we could do that. The requirements for this scrolling effect are complicated enough that we need the flexibility of slivers, which requires the CustomScrollView. The Scaffold doesn't really participate in scrolling directly so I am not sure how we could better integrate these APIs.

@Piinks do you have any thoughts on this?

**Comment by cedvdb** [comment]
> The requirements for this scrolling effect are complicated enough that we need the flexibility of slivers, which requires the CustomScrollView

FWIW, I use a `SliverScaffold` in my apps that takes care of this of abstracting the inner custom scroll view and slivers, maybe something similar could be done.  The only thing annoying with this method is that the slivers are not completely hidden (you have to use a SliverAppBar for example).

Or maybe letting the user do that himself is a better solution because usually only a very light subset of the API is used and the code cannot be simpler. For example here is the `SliverScaffold` used in our current app:

```
class SliverScaffold extends StatelessWidget {
  // ...
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: CustomScrollView(
        slivers: <Widget>[
          appBar,
          SliverToBoxAdapter(child: body),
        ],
      ),
      bottomNavigationBar: bottom,
    );
  }
```

**Comment by Piinks** [comment]
I am not sure what you mean @cedvdb, can you elaborate? 

A SliverAppBar can only be used in a CustomScrollView, and I am not sure what you mean by passing through it.

The Scaffold is not aware of its contents, including the presence of a CustomScrollView. The Scaffold is already a pretty heavily loaded widget, and adding more API surface and complexity to it is likely something we would want to avoid doing.

**Comment by rydmike** [comment]
@darrenaustin @cedvdb @Piinks 
Is it on purpose, or an oversight that text style for `_MediumScrollUnderFlexibleConfig` and `_LargeScrollUnderFlexibleConfig` do not respect and fall through via `AppBarTheme.of(context)`?

They now do this:
```dart
  @override
  TextStyle? get collapsedTextStyle =>
    _textTheme.titleLarge?.apply(color: _colors.onSurface);

  @override
  TextStyle? get expandedTextStyle =>
    _textTheme.headlineSmall?.apply(color: _colors.onSurface);
```

I would expect them to fall through via `AppBarTheme.of(context).foregroundColor` to respect and use themed `AppBar` foreground color, like the vanilla `AppBar` does.

If I for example make an `AppBarTheme` with some blue background and white foreground, it works great on `AppBar`, but themed foreground color is ignored by these new sliver versions `SliverAppBar.medium` and `SliverAppBar.large` using `_MediumScrollUnderFlexibleConfig` and `_LargeScrollUnderFlexibleConfig`, while the themed background is used, as shown here:

Screen Shot 2022-08-23 at 18 52 58

**Comment by HansMuller** [comment]
@darrenaustin - https://github.com/flutter/flutter/pull/103962#issuecomment-1224269768 looks like it covers some legitimate bugs in the new variants; please file issues

**Comment by rydmike** [comment]
@HansMuller and @darrenaustin for info, I opened an issue for it here: https://github.com/flutter/flutter/issues/110951

**Pull request merged**